### PR TITLE
Pipenet revamp

### PR DESCRIPTION
**Pull request by Markolie**
Whoop whoop! I touched pipenet code again and came out alive after losing many hours of my life.

Ports two PR's by Aranclanos (check those for the details), modified to work with our pipenet code:
https://github.com/tgstation/-tg-station/pull/4627
https://github.com/tgstation/-tg-station/pull/5638

Major changes:
* Pipenet code is now 75% less bad.
* Pipes (and pipenets) will now garbage collect properly, meaning less lag during explosions/Lord Singuloth.
* Pipes can now be constructed even when they're not already connected to a pipe.
* An error message will now show up in Dream Daemon when a pipe is in two pipenets.
* Fixes ventcrawling so you can no longer switch between supply/scrubber pipes.
* Adds logging to valves and other atmos equipment.
* Destroyed pipes will now release their gas to the turf.
* The pipe freezer has been removed. Instead, you can now unwrench all atmos machinery, but if the internal pressure is greater than the environment you'll be violently thrown away from the machine. By RemieRichards: https://github.com/tgstation/-tg-station/pull/9867.

### CODE REVIEW

File: code/ATMOSPHERICS/datum_pipeline.dm [hunk 1]
```diff
@@ -1,36 +1,144 @@
...
 	if(air && air.volume)
 		temporarily_store_air()
-		qdel(air)
-
-	return ..()
+	for(var/obj/machinery/atmospherics/pipe/P in members)
+		P.parent = null
+	for(var/obj/machinery/atmospherics/A in other_atmosmch)
+		A.nullifyPipenet(src)
+	return QDEL_HINT_QUEUE
```
Comment (Fox-McCloud) [inline code/ATMOSPHERICS/datum_pipeline.dm]: You'll want to call the parent proc here, somewhere.

File: code/ATMOSPHERICS/components/binary_devices/dp_vent_pump.dm [hunk 2]
```diff
@@ -261,51 +233,26 @@
...
-	user << "<span class='notice'>You begin to unfasten \the [src]...</span>"
-	if (do_after(user, 40, target = src))
-		user.visible_message( \
-			"[user] unfastens \the [src].", \
-			"<span class='notice'>You have unfastened \the [src].</span>", \
-			"You hear a ratchet.")
-		new /obj/item/pipe(loc, make_from=src)
-		qdel(src)
+		
+	return ..()
```
Comment (tigercat2000) [inline code/ATMOSPHERICS/components/binary_devices/dp_vent_pump.dm]: Shouldn't this include arguments in the ..() call?

**Comment by Markolie** [comment]
Reminder to self to disable TESTING.

File: code/ATMOSPHERICS/components/binary_devices/dp_vent_pump.dm (hunk 2 above)
Comment (Markolie) [inline code/ATMOSPHERICS/components/binary_devices/dp_vent_pump.dm]: When no arguments are passed into ..(), it automatically includes all the arguments of its owner proc.

File: code/ATMOSPHERICS/datum_pipeline.dm (hunk 1 above)
Comment (Markolie) [inline code/ATMOSPHERICS/datum_pipeline.dm]: You're right, whoops.

**Comment by Markolie** [comment]
Note to self: fix away missions.

**Comment by tigercat2000** [comment]
> - An error message will now show up in Dream Daemon when a pipe is in two pipenets.

Question; Where is this code stored?

**Comment by Markolie** [comment]
Here: https://github.com/Markolie/Paradise/blob/pipenet/code/ATMOSPHERICS/datum_pipeline.dm#L73

**Comment by tigercat2000** [comment]
Ah, thanks.

**Comment by Fox-McCloud** [comment]
@Aranclanos would be proud.

**Comment by Markolie** [comment]
This is ready to go when Travis agrees. Only downside is that omni devices are currently broken (but unavailable anyway) and aren't really worth fixing.

File: code/ATMOSPHERICS/atmospherics.dm [hunk 3]
```diff
@@ -105,58 +117,116 @@ obj/machinery/atmospherics/proc/check_connect_types_construction(obj/machinery/a
...
+		var/unsafe_wrenching = FALSE
+		var/internal_pressure = int_air.return_pressure()-env_air.return_pressure()
+
+		playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+		user << "<span class='notice'>You begin to unfasten \the [src]...</span>"
+		if (internal_pressure > 2*ONE_ATMOSPHERE)
+			user << "<span class='warning'>As you begin unwrenching \the [src] a gush of air blows in your face... maybe you should reconsider?</span>"
+			unsafe_wrenching = TRUE //Oh dear oh dear
+		
+		if (do_after(user, 40, target = src) && !gcDestroyed)
```
Comment (Fox-McCloud) [inline code/ATMOSPHERICS/atmospherics.dm]: I believe this should be isnull(gcDestroyed)...for whatever reason Krausus made our gcDestroyed different and you pretty much have to do the opposite of this (isValid and a lot of other things are also in line with this); you may want to look into GC code to confirm/deny.

File: code/ATMOSPHERICS/pipes/pipe.dm [hunk 4]
```diff
@@ -1,24 +1,37 @@
...
+
+	var/turf/T = loc
+	for(var/obj/machinery/meter/meter in T)
+		if(meter.target == src)
+			var/obj/item/pipe_meter/PM = new (T)
+			meter.transfer_fingerprints_to(PM)
+			qdel(meter)
+	. = ..()
+
+	if(parent && !parent.gcDestroyed)
```
Comment (Fox-McCloud) [inline code/ATMOSPHERICS/pipes/pipe.dm]: isnull(parent.gcDestroyed) is the correct format here, I believe.

**Comment by Aranclanos** [comment]
I hope you included all following updates made by mso and dunc

**Comment by Markolie** [comment]
Unfortunately enough. This already took me forever to do, and the primary goal of this (initially) was just to make sure pipes/pipenets GC 90% of the time (which this accomplished). Perhaps at a later date those changes can be included.

**Comment by Aranclanos** [comment]
well good luck you gigantic bastard. At that point they were okayish, maybe one error here and there.
Then they were utterly destroyed by someone else and after a lot of screams i got them to fix it. Right now, on tg the pipenets are legit as fuck, with even more cool shit.

**Comment by Fox-McCloud** [comment]
Well, they might not be "legit as fuck" in this PR comparedt o TG,but they're 1000% better than what we have on live.

Currently on live, pipenet datums fail to GC like 99% of the time, and pipes fail to GC 99% of the time, as well---this makes it so pipenet's GC 99% of the time and pipes GC something like 85-90% of the time so....it's a GIGANTIC improvement to us.

**Comment by Aranclanos** [comment]
oh yeah the GC thing hehehehe. Enjoy your singularity NO LAG if this works.
Also being able to wrench pipes down without seeing it disappear in certain cases

**Comment by Markolie** [comment]
Lord Singuloth will be pleased!

**Comment by RemieRichards** [comment]
also enjoy unwrenching pipes smacking you into the wall if they're over-pressured :P

**Comment by Fox-McCloud** [comment]
@PsiOmegaDelta 

You may want to look into this; this retains the unique system that Bay has that allows air distribution and waste pipes to exist on the same while, while dramatically improving pipenet and pipe's overall performance; I know you guys haven't touched your pipenet yet, but plan to--this would be a huge boost for you guys.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

I'm sure that'll generate more than a handful of rather humorous scenarios =p

**Pull request merged**